minigui/README.md

```diff
@@ -53,36 +69,247 @@ A UI for Minigo
 
 1. Run `./fetch-and-run.sh`
 
-1. open localhost:5001 (or whatever value you used for $MINIGUI_PORT).
+1. open localhost:5001 (or whatever value you used for $MINIGUI\_PORT).
 
 1. The buttons in the upper right that say 'Human' can be toggled to set which
    color Minigo will play.
 
-**C++ Engine**
+## Minigui modes
 
-1. By default, Minigui will use the Python Minigo engine. You can use the C++
-   engine by first compiling it (see the
-   [README](https://github.com/tensorflow/minigo/tree/master/cc/README.md)),
-   then passing `--engine=<tf,tpu,lite,trt>` when starting `minigui/serve.py`.
-   You will need to pass a frozen GraphDef proto as the `--model` command line
-   argument instead of the saved parameter data that the Python backend
-   requires.
+Minigui has various modes of operation.
 
-   **Note:** Compiling tensorflow from scratch can take 2+ hours if your
-   machine is not terribly beefy, So you might want to kick off the build and
-   get a coffee.
+### Study mode
 
-1. Before running, you'll need to freeze a model so the C++ job can consume it.
-   This assumes a converted model from above.
+Study mode allows you to load previously recorded SGF games, explore variations
+and explore what the engine thinks of each position.
 
-    ```shell
-    python main.py freeze-graph $MODEL_DIR/$MODEL.converted
-    ```
+**Minigui's study mode currently requires the C++ Minigo engine for full
+functionality.**
 
-1. Here's an example, note that we specify the board size when compiling the
-   C++ engine binary, not when running the Minigui server:
+### Vs mode
 
-    ```shell
-    bazel build -c opt --define=tf=1 cc:main
-    python minigui/serve.py --port=8888 --model=$MODEL_DIR/$MODEL.converted.pb --engine=tf
-    ```
+A mode that plays two models or engines against each other and displays the
+variations that each player is considering in real time.
+
+### Demo mode
+
+The original demo built for GTC.
+
+### Lightweight Demo mode
+
+A lightweight demo mode built as experiment for running Minigo on a Raspberry
+Pi.
+
+### Kiosk mode
+
+A self-play only version of the demo mode.
+
+
+## Technical discussion
+
+A technical discussion of the features that a Go engine must implement to
+support Minigui now follows.
+
+### Communicating game state
+
+#### Position messages
+
+The Minigui frontend doesn't understand the rules of Go. Consequently, every
+time a move is played or a new game is started, the engine must communicate the
+game state to the frontend. This is done by printing a JSON object on a single
+line to `stderr` with the prefix `mg-position:`.
+
+The JSON object is made up of the following fields (a `?` denotes an optional
+field, other fields are required):
+
+ - `id: string`: an ID that uniquely identifies the current board
+   position.
```

Comment:
"(we use can be syzgygygygy hash)" (if that's true)

Reply:
actually, I realized a couple of days ago that the Zobrist hash of the stones won't work for study mode because multiple variations could lead to the same board state, so it's no longer a unique ID